File: BOJ/silver/4949.py
# 균형잡힌 세상

answer = []
sign = {')': '(', ']': '['}

while True:
    str_ = input()
    if str_ == '.':
        break
    prev = []
    check = True
    for i in str_:
        if i in sign.values():
            prev.append(i)
        elif i in sign.keys():
            if prev:
                if prev[-1] == sign[i]:
                    prev.pop()
                else:
                    check = False
            else:
                check = False
                break
    if prev:
        answer.append('no')
    else:
        if check:
            answer.append('yes')
        else:
            answer.append('no')

for i in answer:
    print(i)

File: BOJ/silver/4949.js
// 4949 - 균형잡힌 세상

const fs = require("fs");
const file = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const inputs = fs.readFileSync(file).toString().split("\n");

const bracketPairs = {
  "]": "[",
  ")": "(",
};
const openBrackets = Object.values(bracketPairs);
const closeBrackets = Object.keys(bracketPairs);
const answer = [];

let stack;
let isBreak;

for (const input of inputs) {
  if (input === ".") break;

  stack = [];
  isBreak = false;

  for (const char of input) {
    if (openBrackets.includes(char)) {
      stack.push(char);
    } else if (closeBrackets.includes(char)) {
      if (stack[stack.length - 1] === bracketPairs[char]) {
        stack.pop();
      } else {
        answer.push("no");
        isBreak = true;
        break;
      }
    }
  }

  if (!isBreak) {
    if (stack.length === 0) {
      answer.push("yes");
    } else {
      answer.push("no");
    }
  }
}

console.log(answer.join("\n"));
